GH-9740:  Add plugin methods to plugin API

hanzei opened the pull request:
#### Summary
This PR adds a couple methods to interact this plugins to the plugin API 

#### Ticket Link
Fixes #9740

jwilander commented on app/plugin_api.go:
```diff
@@ -422,6 +422,30 @@ func (api *PluginAPI) GetEmojiImage(emojiId string) ([]byte, string, *model.AppE
+	return api.app.GetPlugins()
+}
+
+func (api *PluginAPI) GetPluginStatuses() (model.PluginStatuses, *model.AppError) {
```
We could but since it's just type alias of `[]*PluginStatus` we don't really need to as slices are already pointers to begin with

crspeller replied:
Not sure we should use it here. I think it would be more clear if we just returned `[]*PluginStatus`.

jwilander commented on app/plugin_api.go:
```diff
@@ -422,6 +422,30 @@ func (api *PluginAPI) GetEmojiImage(emojiId string) ([]byte, string, *model.AppE
 
+// Plugin Section
+
+func (api *PluginAPI) GetPlugins() (*model.PluginsResponse, *model.AppError) {
```
Hmm I wonder if it would be better to just return a slice of plugin manifests here rather than use `model.PluginsResponse`. If you want plugin statuses you can use `GetPluginStatuses` instead.

Another consideration would be we have this return a list of all plugin manifests, remove `GetPluginStatuses` and just add `GetPluginStatus` so that a plugin can pull that status for only the plugins it cares about. Thoughts @hanzei @crspeller ?

crspeller replied:
Ya I think it would be better to go with @jwilander suggestion.

hanzei commented on app/plugin_api.go:
```diff
@@ -426,6 +426,41 @@ func (api *PluginAPI) GetEmojiImage(emojiId string) ([]byte, string, *model.AppE
+// Plugin Section
+
+func (api *PluginAPI) GetPlugins() ([]*model.Manifest, *model.AppError) {
+	plugins, err := api.app.GetPlugins()
```
Can this logic stay there or should it be moved to `app/plugins.go`?

jwilander replied:
It can stay here but I'd suggest adding a small test since it's more than just 1-2 lines

hanzei replied:
Is there an easier way to set up plugins in a test environment then
```go
th := Setup().InitBasic()
defer th.TearDown()
setupPluginApiTest(t, "InsertGoCodeHere", "InsertManifestHere", "id", th.App)
``` 
? Is there some kind of sample Plugin which is reused in multiple tests?

hanzei replied:
@jwilander Friendly ping ^

crspeller replied:
No, I don't think there are any tests that do that at the moment. If you can see a better way to test these, then feel free to propose something.

jwilander replied:
Sorry @hanzei for the late response, basically what @crspeller said

hanzei commented on app/plugin_statuses.go:
```diff
@@ -9,6 +9,27 @@ import (
+			return status, nil
+		}
+	}
+	return nil, model.NewAppError("GetPluginStatus", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
```
AppErrors always confuse me. Can you guys just tell me what to return?

hanzei commented on app/plugin_statuses.go:
```diff
@@ -9,6 +9,27 @@ import (
+			return status, nil
+		}
+	}
+	return nil, model.NewAppError("GetPluginStatus", "app.plugin.not_installed.app_error", nil, "", http.StatusBadRequest)
```
AppErrors always confuse me. Is this correct?

jwilander replied:
Looks good 👍

hanzei commented:
I added a test to `app/plugin_api_test.go`. I know it is hacky but a prettier solution would require a lot of refactoring - which should definitely be done but not in this PR.

I added some checks to `setupPluginApiTest()`.

@jwilander @crspeller Please review again.
